client/core/index.ts

```diff
@@ -0,0 +1,240 @@
+import {
+  CosmWasmClient,
+  SigningCosmWasmClient,
+} from '@cosmjs/cosmwasm-stargate'
+import { WalletData } from './wallet'
+import connectCosmWasmClient from './cosmwasm'
+import { ChainData } from './chain'
+import { MeshConsumerQueryClient } from 'codegen/MeshConsumer.client'
+import {
+  MeshLockupClient,
+  MeshLockupQueryClient,
+} from 'codegen/MeshLockup.client'
+import {
+  MeshProviderClient,
+  MeshProviderQueryClient,
+} from 'codegen/MeshProvider.client'
+import {
+  MeshSlasherClient,
+  MeshSlasherQueryClient,
+} from 'codegen/MeshSlasher.client'
+import {
+  MetaStakingClient,
+  MetaStakingQueryClient,
+} from 'codegen/MetaStaking.client'
+
+import { junoContracts, osmoContracts } from 'config'
+
+export interface MeshClientConstructor {
+  wallet: WalletData | null
+  chain: ChainData
+  signingCosmWasmClient: SigningCosmWasmClient | null
+}
+
+export type ChainName = 'junotestnet' | 'osmosistestnet'
+
+export class MeshClient {
+  private _cosmWasmClient: CosmWasmClient | null = null
+  public signingCosmWasmClient: SigningCosmWasmClient | null = null
+
+  private _meshConsumerClient: MeshConsumerQueryClient | null = null
+
+  private _meshLockupClient: MeshLockupQueryClient | null = null
+  public signingMeshLockupClient: MeshLockupClient | null = null
+
+  private _meshProviderClient: MeshProviderQueryClient | null = null
+  public signingMeshProviderClient: MeshProviderClient | null = null
+
+  private _meshSlasherClient: MeshSlasherQueryClient | null = null
+  public signingMeshSlasherClient: MeshSlasherClient | null = null
+
+  private _metaStakingClient: MetaStakingQueryClient | null = null
+  public signingMetaStakingClient: MetaStakingClient | null = null
+
+  public chain: ChainData
+
+  private _wallet: WalletData | null = null
+
+  constructor({ wallet, chain, signingCosmWasmClient }: MeshClientConstructor) {
+    this._wallet = wallet
+    this.chain = chain
+    this.signingCosmWasmClient = signingCosmWasmClient
+  }
+
+  public async connect(chain: ChainName) {
+    // CosmWasm client already exists!
+    if (this._cosmWasmClient) {
+      return
+    }
+
+    // Create CosmWasm client
+    this._cosmWasmClient = await connectCosmWasmClient(this.chain.rpc)
+
+    // Create all contract clients
+    await this.createContractClients(chain)
+  }
+
+  // Asynchronously init every client at the same time
+  private async createContractClients(chain: ChainName) {
+    if (!chain) return
+    const contracts = {
+      junotestnet: [
+        this.createMeshConsumerClient,
```

Comment:
We wait for the init to be completed here:
```
const completed = await Promise.all(initClients)
```


Function won't complete until all of them have been initialized. IMO this is more efficient that individually awaiting every one of them. Also want to avoid initializing clients right before they are needed, will take some time off of queries and will also save some resources. Every MB of memory matters!